VSDiagnostics/VSDiagnostics/VSDiagnostics/Diagnostics/General/FlagsEnumValuesAreNotPowersOfTwo/FlagsEnumValuesAreNotPowersOfTwoAnalyzer.cs

```diff
@@ -0,0 +1,93 @@
+﻿using System;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace VSDiagnostics.Diagnostics.General.FlagsEnumValuesAreNotPowersOfTwo
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public class FlagsEnumValuesAreNotPowersOfTwoAnalyzer : DiagnosticAnalyzer
+    {
+        private const string DiagnosticId = nameof(FlagsEnumValuesAreNotPowersOfTwoAnalyzer);
+        private const DiagnosticSeverity Severity = DiagnosticSeverity.Error;
+
+        private static readonly string Category = VSDiagnosticsResources.GeneralCategory;
+        private static readonly string Message = VSDiagnosticsResources.FlagsEnumValuesAreNotPowersOfTwoAnalyzerMessage;
+        private static readonly string Title = VSDiagnosticsResources.FlagsEnumValuesAreNotPowersOfTwoAnalyzerTitle;
+
+        internal static DiagnosticDescriptor Rule => new DiagnosticDescriptor(DiagnosticId, Title, Message, Category, Severity, true);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.RegisterSyntaxNodeAction(AnalyzeSymbol, SyntaxKind.EnumDeclaration);
+        }
+
+        private void AnalyzeSymbol(SyntaxNodeAnalysisContext context)
+        {
+            var declarationExpression = context.Node as EnumDeclarationSyntax;
+            if (declarationExpression == null)
+            {
+                return;
+            }
+
+            if (!declarationExpression.AttributeLists.Any(
+                    a => a.Attributes.Any(
+                        t => context.SemanticModel.GetSymbolInfo(t).Symbol.ContainingType.MetadataName == typeof(FlagsAttribute).Name)))
+            {
+                return;
+            }
+
+            var enumMemberDeclarations = declarationExpression.ChildNodes().OfType<EnumMemberDeclarationSyntax>().ToList();
+            var values = enumMemberDeclarations.Select(member => member.EqualsValue);
+
+            foreach (var equalsValue in values)
+            {
+                if (equalsValue == null)
+                {
+                    var enumName = context.SemanticModel.GetDeclaredSymbol(declarationExpression).Name;
+                    context.ReportDiagnostic(Diagnostic.Create(Rule, declarationExpression.GetLocation(), enumName));
+                    return;
+                }
+
+                LiteralExpressionSyntax valueExpression = null;
+
+                if (equalsValue.Value is LiteralExpressionSyntax)
+                {
+                    valueExpression = (LiteralExpressionSyntax) equalsValue.Value;
+                }
+                if (equalsValue.Value is PrefixUnaryExpressionSyntax)
```

Comment:
Can you add some general commentary in certain places to make clear what scenario you're checking for at that particular point?